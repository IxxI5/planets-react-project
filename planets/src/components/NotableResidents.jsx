import { FaMale } from "react-icons/fa";
import { FaFemale } from "react-icons/fa";
import { FaGenderless } from "react-icons/fa";

/**
 * Displays the notable residents of a planet
 * @param {string} planet: single object
 */
export default function NotableResidents({ planet }) {
  return (
    <td>
      {planet.residents.map((resident, i) => {
        let genderIcon = <FaGenderless title="Gender N/A" />;

        // gender to FA icon converter
        switch (resident.gender) {
          case "male":
            genderIcon = <FaMale title="Male" />;
            break;
          case "female":
            genderIcon = <FaFemale title="Female" />;
            break;
          default:
            genderIcon = <FaGenderless title="Gender N/A" />;
            break;
        }

        let person = Object.values(resident);
        person.pop(); // remove gender text value

        return (
          <p key={i}>
            {genderIcon} {person.join(", ")}
          </p>
        );
      })}
    </td>
  );
}
