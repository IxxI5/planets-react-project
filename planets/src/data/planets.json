{
  "planets": [
    {
      "id": "0",
      "name": "Tatooine",
      "diameter": "10465",
      "gravity": "1 standard",
      "terrain": "desert",
      "population": "200000",
      "residents": [
        {
          "name": "Luke Skywalker",
          "height": "172",
          "birth_year": "19BBY",
          "gender": "male"
        },
        {
          "name": "C-3PO",
          "height": "167",
          "birth_year": "112BBY",
          "gender": "n/a"
        }
      ]
    },
    {
      "id": "1",
      "name": "Alderaan",
      "diameter": "12500",
      "gravity": "1 standard",
      "terrain": "grasslands, mountains",
      "population": "2000000000",
      "residents": [
        {
          "name": "Leia Organa",
          "height": "150",
          "birth_year": "19BBY",
          "gender": "female"
        },
        {
          "name": "Bail Prestor Organa",
          "height": "191",
          "birth_year": "67BBY",
          "gender": "male"
        },
        {
          "name": "Raymus Antilles",
          "height": "188",
          "birth_year": "unknown",
          "gender": "male"
        }
      ]
    },
    {
      "id": "2",
      "name": "Yavin IV",
      "diameter": "10200",
      "gravity": "1 standard",
      "terrain": "jungle, rainforests",
      "population": "1000",
      "residents": []
    },
    {
      "id": "3",
      "name": "Hoth",
      "diameter": "7200",
      "gravity": "1.1 standard",
      "terrain": "tundra, ice caves, mountain ranges",
      "population": "unknown",
      "residents": []
    },
    {
      "id": "4",
      "name": "Dagobah",
      "diameter": "8900",
      "gravity": "N/A",
      "terrain": "swamp, jungles",
      "population": "unknown",
      "residents": []
    }
  ]
}
