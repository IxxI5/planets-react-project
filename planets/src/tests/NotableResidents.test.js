import "@testing-library/jest-dom"; // pre-requisite: npm i jest-dom
import renderer from "react-test-renderer"; // pre-requisite: npm i react-test-renderer
import { FaMale } from "react-icons/fa";
import { FaGenderless } from "react-icons/fa";
import NotableResidents from "../components/NotableResidents";
import { planets } from "../data/planets.json";

// test data
const planet = planets[0];
delete planet.id;

// data fromatting => remove gender text
let person1 = Object.values(planet.residents[0]);
let person2 = Object.values(planet.residents[1]);
person1.pop(); // remove gender text value
person2.pop(); // remove gender text value

/* helping component */
function TD() {
  // Render the component to resolve the Error "Functions are not valid as a React child"
  //   Case 3: https://bobbyhadz.com/blog/react-functions-are-not-valid-as-react-child
  const Ps = () => {
    return (
      <>
        <p>
          <FaMale /> {person1.join(", ")}
        </p>
        <p>
          <FaGenderless /> {person2.join(", ")}
        </p>
      </>
    );
  };

  return <Ps />;
}

/* rendering tests using MatchSnapshot */
test(`(0) | create a reference snapshot`, () => {
  /* Arrange */
  /* render components plus Act -> JSON*/
  const ref = renderer.create(<NotableResidents planet={planet} />).toJSON();

  /* Assert */
  expect(ref).toMatchSnapshot();
});

/* compare rendered helping component with the snapshot */
test(`(1) | compare helping component with the reference snapshot`, () => {
  /* Arrange */
  /* render components plus Act -> JSON*/
  const comp = renderer.create(<TD />).toJSON();

  /* Assert */
  expect(comp).toMatchSnapshot();
});
