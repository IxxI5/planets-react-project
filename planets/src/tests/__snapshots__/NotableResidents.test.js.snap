// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`(0) | create a reference snapshot 1`] = `
<td>
  <p>
    <svg
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      strokeWidth="0"
      style={
        Object {
          "color": undefined,
        }
      }
      viewBox="0 0 192 512"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <title>
        Male
      </title>
      <path
        d="M96 0c35.346 0 64 28.654 64 64s-28.654 64-64 64-64-28.654-64-64S60.654 0 96 0m48 144h-11.36c-22.711 10.443-49.59 10.894-73.28 0H48c-26.51 0-48 21.49-48 48v136c0 13.255 10.745 24 24 24h16v136c0 13.255 10.745 24 24 24h64c13.255 0 24-10.745 24-24V352h16c13.255 0 24-10.745 24-24V192c0-26.51-21.49-48-48-48z"
      />
    </svg>
     
    Luke Skywalker, 172, 19BBY
  </p>
  <p>
    <svg
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      strokeWidth="0"
      style={
        Object {
          "color": undefined,
        }
      }
      viewBox="0 0 288 512"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <title>
        Gender N/A
      </title>
      <path
        d="M144 176c44.1 0 80 35.9 80 80s-35.9 80-80 80-80-35.9-80-80 35.9-80 80-80m0-64C64.5 112 0 176.5 0 256s64.5 144 144 144 144-64.5 144-144-64.5-144-144-144z"
      />
    </svg>
     
    C-3PO, 167, 112BBY
  </p>
</td>
`;

exports[`(1) | compare helping component with the reference snapshot 1`] = `
Array [
  <p>
    <svg
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      strokeWidth="0"
      style={
        Object {
          "color": undefined,
        }
      }
      viewBox="0 0 192 512"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M96 0c35.346 0 64 28.654 64 64s-28.654 64-64 64-64-28.654-64-64S60.654 0 96 0m48 144h-11.36c-22.711 10.443-49.59 10.894-73.28 0H48c-26.51 0-48 21.49-48 48v136c0 13.255 10.745 24 24 24h16v136c0 13.255 10.745 24 24 24h64c13.255 0 24-10.745 24-24V352h16c13.255 0 24-10.745 24-24V192c0-26.51-21.49-48-48-48z"
      />
    </svg>
     
    Luke Skywalker, 172, 19BBY
  </p>,
  <p>
    <svg
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      strokeWidth="0"
      style={
        Object {
          "color": undefined,
        }
      }
      viewBox="0 0 288 512"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M144 176c44.1 0 80 35.9 80 80s-35.9 80-80 80-80-35.9-80-80 35.9-80 80-80m0-64C64.5 112 0 176.5 0 256s64.5 144 144 144 144-64.5 144-144-64.5-144-144-144z"
      />
    </svg>
     
    C-3PO, 167, 112BBY
  </p>,
]
`;
